### ISSUE TITLE
function to get list of *occurrences* of free variables

### ISSUE THREAD

**Issue by jonsterling** [RedPRL/sml-redprl#110]
(i.e. with their annotations)

**Comment by jonsterling** [comment]
This will be useful for providing better error messages; when a variable occurs unbound, we could underline the exact spot in Emacs.

This ticket should probably be addressed in [sml-typed-abts](https://github.com/RedPRL/sml-typed-abts/), by adding operations to the `ABT` signature.

**Comment by ak3n** [comment]
What is purpose of [`varctx`](https://github.com/RedPRL/sml-typed-abts/blob/master/src/core/abt.fun#L206)?

**Comment by jonsterling** [comment]
@ak3n This gives you the collection of free variables in a term, along with their sorts.

**Comment by wilcoxjay** [comment]
I'd like to take a crack at this one. Here's what I'm thinking, based on the proposal above. Add operations
```
metaOccurrences : abt -> annotation Metavar.ctx
varOccurrences : abt -> annotation Var.ctx
symOccurrences : abt -> annotation Sym.ctx
```
to the ABT signature in `sml-typed-abts`.

Then, use these operations to get better error messages in [checkScope](https://github.com/RedPRL/sml-redprl/blob/93684fa8ad3dc16b3e298386e857fa822ce7a36e/src/redprl/signature.sml#L243). In two cases, that code only cares about the key but not the value in the mapping (ie, the variable, but not its sort/valence). In those cases, we can just replace `*ctx` with `*Occurrences`. But in the case of symbols, that code *does* care about the value, and so we need to call both of them. 

An alternate design would be to change the existing `*ctx` operations to return something like `(sort * annotation) Var.ctx`. This has the dual downside of forcing every client to explicitly ignore position information when they don't need it. I'm also not sure how this alternate design would affect the internals of `sml-typed-abts`, which makes several uses of the `*ctx` operations.

**Comment by jonsterling** [comment]
I like your first proposal... Though it would probably need to be something like `metaOccurrences : abt -> annotation list Metavar.ctx` or something along those lines, since a variable can occur multiple times (and the contexts only include a variable once).

**Comment by wilcoxjay** [comment]
Oh, right, that makes sense.

**Comment by wilcoxjay** [comment]
This is closed by #150.

I'm going to open a similar issue wrt metavariables
